### PR TITLE
[RFC] Introduce new pattern to avoid duplication of core services to extensions

### PR DESCRIPTION
**Pull request by cramforce**

### CODE REVIEW

**Comment by cramforce** [comment]
CC @dvoytenko @erwinmombay

**Comment by dvoytenko** [comment]
I think this is fine. But I still wanted to @erwinmombay to comment on how easy it'd be to just do it at compilation time?

**Comment by erwinmombay** [comment]
@dvoytenko can you explain again what we needed to do? just creating the installer code in a separate file?

**Comment by cramforce** [comment]
Explicit might be better than implicit (through compiler magic) here.

**Comment by dvoytenko** [comment]
Normally, the service code looks like this:
```
class History {
  // implementation
}

/**
 * @param {!Window} window
 * @return {!History}
 */
export function historyFor(window) {
  return getService(window, 'history', () => {
    return new History(window);
  });
};
```

We'd like to turn this to the following when compiling an extension:
```
/**
 * @param {!Window} window
 * @return {!History}
 */
export function historyFor(window) {
  return getService(window, 'history');
};
```

In short, we throw away all implementation details and keep this `serviceFor` method and remove the factory from it. This way, an extension would not ever compile redundant details of a service and we have no risk of an extension accidentally either creating a service before main runtime.

Malte is suggesting we do this manually, but putting separate `serviceFor` in a separate file. But my question is whether we can do this automatically.

**Comment by cramforce** [comment]
If we use magic the important part is to also remove all the imports!

**Comment by dvoytenko** [comment]
Well, I'm leaning toward the manual solution as well.

**Comment by erwinmombay** [comment]
right, so if we wanted to do it magically, the annotation would actually have to be on the `import` statement in the extension so that:

1. we can remove the import statement.
2. generate the `serviceFor` pattern embedded into the extension.

File: src/history.js [hunk 1]
```diff
@@ -14,647 +14,11 @@
...
-    binding = new HistoryBindingVirtual_(viewer);
-  } else {
-    binding = new HistoryBindingNatural_(window);
-  }
-  return new History(binding);
-};
-
-
 /**
+ * Gets service implemented in service/history-impl.
```
Comment (dvoytenko) [inline src/history.js]: "Returns service"

File: src/service/history-impl.js [hunk 2]
```diff
@@ -0,0 +1,659 @@
+/**
```
Comment (dvoytenko) [inline src/service/history-impl.js]: Weird. Is it not catching the "move" op? Would be good to preserve the revisions...

**Comment by erwinmombay** [comment]
as far as implementation, that would be really easy to do. just depends on how comfortable we are with magic annotations :smile_cat:

File: src/service/history-impl.js (hunk 2 above)
Comment (cramforce) [inline src/service/history-impl.js]: Unfortunately git does not have a way to move a file and keep something in the original place in the same commit.

File: src/history.js (hunk 1 above)
Comment (cramforce) [inline src/history.js]: Done

File: src/service/history-impl.js (hunk 2 above)
Comment (cramforce) [inline src/service/history-impl.js]: Correction. It does auto detect the copy, but apparently github doesn't show it.

Comment (dvoytenko) [inline src/service/history-impl.js]: Ok. Good.

**Comment by dvoytenko** [comment]
LGTM

**Pull request merged**